### PR TITLE
Get most services out of the default interface 

### PR DESCRIPTION
**Pull request by ktsakalozos**
Ports still used on the default interface:

Port | Service | Access Restrictions
--- | --- | ---
16443 | API server | SSL encrypted. Clients need to present a valid password from a [Static Password File](https://kubernetes.io/docs/reference/access-authn-authz/authentication/#authentication-strategies).
10250 | kubelet | Anonymous authentication is disabled. [X509 client certificate](https://kubernetes.io/docs/reference/command-line-tools-reference/kubelet-authentication-authorization/) is required.
10255 | kubelet | Read only port for the Kubelet.
random | kube-proxy | One random port per hosted service is opened as we use `--proxy-mode=userspace` for compatibility reasons.

If you remove `--proxy-mode` from `/var/snap/microk8s/current/args/kube-proxy` and `sudo systemctl restart snap.microk8s.daemon-proxy` kube-proxy will stop exposing the cluster hosted services.

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -13,7 +13,7 @@ snap install microk8s --classic --beta
 ```
 
 > At this time microk8s is an early beta, while this should be safe to install please beware.
-> In order to install microk8s make sure port 8080 is not used.
+> There are a few ports that need to be open for microk8s to install successfully. 8080 and 6443 are the most common ones. Please go though the [list of used ports](docs/ports.md) and the security implications involved in installing microk8s.
```
Comment (tvansteenburgh) [inline README.md]: s/open/available/

File: docs/ports.md [hunk 2]
```diff
@@ -0,0 +1,58 @@
+# Services Exposed and Ports Used
+
+For now microk8s is meant to be used for local development thus certain security issues are not addressed at this point. Here we present the ports and sockets each service uses as well as the default authorisation and authentication configuration.
+
+Services can be placed in two groups based on the network interface they are bind to. Services binding to the localhost interface are only available from within the host and we take no action to protect them. Services binding to the default host interface are available from outside the host and thus we enforce some form of access restrictions. The ports used by both types of services need to be free so that microk8s starts successfully.
```
Comment (tvansteenburgh) [inline docs/ports.md]: "interface they are bound to" or "interface they bind to"

File: docs/ports.md [hunk 3]
```diff
@@ -0,0 +1,58 @@
...
+
+### Services Binging to the Default Host Interface
```
Comment (tvansteenburgh) [inline docs/ports.md]: s/Binging/Binding/

File: docs/ports.md [hunk 4]
```diff
@@ -0,0 +1,58 @@
...
+--- | --- | ---
+6443 | API server | SSL encrypted. Clients need to present a valid password from a [Static Password File](https://kubernetes.io/docs/reference/access-authn-authz/authentication/#authentication-strategies).
+10250 | kubelet | Anonymous authentication is disabled. [X509 client certificate](https://kubernetes.io/docs/reference/command-line-tools-reference/kubelet-authentication-authorization/) is required.
+10255 | kubelet | Read only port for the Kubelet.
+random | kube-proxy | One random port per hosted service is opened as we use `--proxy-mode=userspace` for compatibility reasons.
+
+If you remove `--proxy-mode` from `/var/snap/microk8s/current/args/kube-proxy` and `sudo systemctl restart snap.microk8s.daemon-proxy` kube-proxy will stop exposing the cluster hosted services.
+
+
+### Services Binging to the localhost Interface
```
Comment (tvansteenburgh) [inline docs/ports.md]: s/Binging/Binding/

File: docs/ports.md [hunk 5]
```diff
@@ -0,0 +1,58 @@
...
+
+Service | Socket
+--- | ---
+docker | unix:///var/snap/microk8s/current/docker.sock
+etcd | unix://etcd.socket:2379
+
+
+## Authentication and Authorization
+
+Upon a new deployment microk8s creates a new CA, a signed server certificate and a service account key file. These files are stored under `/var/microk8s/current/certs`. Kubelet an the API server are aware of the same CA and so the signed server certificate is used by the API server to authenticate with kubelet (`--kubelet-client-certificate`). Clients talking to the secure port of the API server (`6443`) have to also be aware of the CA (`certificate-authority-data` in user kubeconfig).
```
Comment (tvansteenburgh) [inline docs/ports.md]: Kubelet *and* the API server...

File: snapcraft.yaml.orig [hunk 6]
```diff
@@ -0,0 +1,200 @@
+name: microk8s
```
Comment (tvansteenburgh) [inline snapcraft.yaml.orig]: Is this file (snapcraft.yaml.orig) supposed to be here?

**Comment by ktsakalozos** [comment]
> I would like to test this PR

@nilbot The channel `1.13/edge/security-testing` was a temporary one. I had to re-upload a new snap package. You can now try:
```
sudo snap install microk8s --classic --channel=1.13/edge/security-testing
```

**Comment by ktsakalozos** [comment]
Closing this in favor of https://github.com/ubuntu/microk8s/pull/323

**Pull request closed without merging**